src/witness_generator/machines/block_machine.rs

```diff
@@ -0,0 +1,551 @@
+    degree: DegreeType,
+    /// Cache that states the order in which to evaluate identities
+    /// to make progress most quickly.
+    processing_sequence_cache: ProcessingSequenceCache,
+}
+
+impl BlockMachine {
+    pub fn try_new(
+        fixed_data: &FixedData,
+        connecting_identities: &[&Identity],
+        identities: &[&Identity],
+        witness_names: &HashSet<&str>,
+        global_bit_constraints: &BTreeMap<&str, BitConstraint>,
+    ) -> Option<Box<Self>> {
+        for id in connecting_identities {
+            if let Some(sel) = &id.right.selector {
+                // TODO we should check that the other constraints/fixed columns are also periodic.
+                if let Some((selector, period)) = is_boolean_periodic_selector(sel, fixed_data) {
+                    let mut machine = BlockMachine {
+                        block_size: period,
+                        selector,
+                        identities: identities.iter().map(|&i| i.clone()).collect(),
+                        data: witness_names
+                            .iter()
+                            .map(|n| (fixed_data.witness_ids[n], vec![]))
+                            .collect(),
+                        row: 0,
+                        bit_constraints: Default::default(),
+                        global_bit_constraints: global_bit_constraints
+                            .iter()
+                            .filter_map(|(n, c)| {
+                                fixed_data.witness_ids.get(n).map(|n| (*n, c.clone()))
+                            })
+                            .collect(),
+                        witness_count: fixed_data.witness_cols.len(),
+                        degree: fixed_data.degree,
+                        processing_sequence_cache: ProcessingSequenceCache::new(
+                            period,
+                            identities.len(),
+                        ),
+                    };
+                    // Append a block so that we do not have to deal with wrap-around
+                    // when storing machine witness data.
+                    machine.append_new_block();
+
+                    return Some(Box::new(machine));
+                }
+            }
+        }
+
+        None
+    }
+}
+
+/// Check if `expr` is a reference to a function of the form
+/// f(i) { if (i + 1) % k == 0 { 1 } else { 0 } }
+/// for some k >= 2
+/// TODO we could make this more generic and only detect the period
+/// but do not enforce the offset.
+fn is_boolean_periodic_selector(
+    expr: &Expression,
+    fixed_data: &FixedData,
+) -> Option<(String, usize)> {
+    let poly = is_simple_poly(expr)?;
+
+    let values = fixed_data.fixed_cols.get(poly)?;
+
+    let period = 1 + values.iter().position(|v| *v == 1.into())?;
+    if period == 1 {
+        return None;
+    }
+    for (i, v) in values.iter().enumerate() {
+        if *v
+            != if (i + 1) % period == 0 {
```

Comment:
Sorry I'm struggling with this syntax 🤔

Reply:
Oh now I see it. The formatter makes it look confusing, it could be more readable with an intermediate variable.

Reply:
Also more generally 
```rust
values.iter().enumerate().all(|(i, v)| ...).then_some((poly.to_string(), period))
```